coderd/coderdtest/oidctest/idp.go

```diff
@@ -0,0 +1,768 @@
+	t.Helper()
+
+	if f.hookAuthenticateClient != nil {
+		return f.hookAuthenticateClient(t, req)
+	}
+
+	data, _ := io.ReadAll(req.Body)
+	values, err := url.ParseQuery(string(data))
+	if !assert.NoError(t, err, "parse token request values") {
+		return nil, xerrors.New("invalid token request")
+	}
+
+	if !assert.Equal(t, f.clientID, values.Get("client_id"), "client_id mismatch") {
+		return nil, xerrors.New("client_id mismatch")
+	}
+
+	if !assert.Equal(t, f.clientSecret, values.Get("client_secret"), "client_secret mismatch") {
+		return nil, xerrors.New("client_secret mismatch")
+	}
+
+	return values, nil
+}
+
+// encodeClaims is a helper func to convert claims to a valid JWT.
+func (f *FakeIDP) encodeClaims(t testing.TB, claims jwt.MapClaims) string {
+	t.Helper()
+
+	if _, ok := claims["exp"]; !ok {
+		claims["exp"] = time.Now().Add(time.Hour).UnixMilli()
+	}
+
+	if _, ok := claims["aud"]; !ok {
+		claims["aud"] = f.clientID
+	}
+
+	if _, ok := claims["iss"]; !ok {
+		claims["iss"] = f.issuer
+	}
+
+	signed, err := jwt.NewWithClaims(jwt.SigningMethodRS256, claims).SignedString(f.key)
+	require.NoError(t, err)
+
+	return signed
+}
+
+// httpHandler is the IDP http server.
+func (f *FakeIDP) httpHandler(t testing.TB) http.Handler {
+	t.Helper()
+
+	mux := chi.NewMux()
+	// This endpoint is required to initialize the OIDC provider.
+	// It is used to get the OIDC configuration.
+	mux.Get("/.well-known/openid-configuration", func(rw http.ResponseWriter, r *http.Request) {
+		_ = json.NewEncoder(rw).Encode(f.provider)
+	})
+
+	// Authorize is called when the user is redirected to the IDP to login.
+	// This is the browser hitting the IDP and the user logging into Google or
+	// w/e and clicking "Allow". They will be redirected back to the redirect
+	// when this is done.
+	mux.Handle(authorizePath, http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		f.logger.Info(r.Context(), "HTTP Call Authorize", slog.F("url", string(r.URL.String())))
+
+		clientID := r.URL.Query().Get("client_id")
+		if clientID != f.clientID {
+			t.Errorf("unexpected client_id %q", clientID)
+			http.Error(rw, "invalid client_id", http.StatusBadRequest)
+		}
+
+		redirectURI := r.URL.Query().Get("redirect_uri")
+		state := r.URL.Query().Get("state")
+
+		scope := r.URL.Query().Get("scope")
+		_ = scope
+
+		responseType := r.URL.Query().Get("response_type")
+		switch responseType {
+		case "code":
+		case "token":
+			t.Errorf("response_type %q not supported", responseType)
+			http.Error(rw, "invalid response_type", http.StatusBadRequest)
+			return
+		default:
+			t.Errorf("unexpected response_type %q", responseType)
+			http.Error(rw, "invalid response_type", http.StatusBadRequest)
+			return
+		}
+
+		ru, err := url.Parse(redirectURI)
+		if err != nil {
+			t.Errorf("invalid redirect_uri %q", redirectURI)
+			http.Error(rw, "invalid redirect_uri", http.StatusBadRequest)
+			return
+		}
```

Comment:
Super smart!

I hope this can evolve and people can add these things as we encounter them. I will add this one, but some knobs I am going to refrain from adding until we write the tests that require them.